tree-sitter/src/TreeSitter/Unmarshal.hs

```diff
@@ -53,75 +66,110 @@ parseByteString language bytestring = withParser language $ \ parser -> withPars
   else
     withRootNode treePtr $ \ rootPtr ->
       withCursor (castPtr rootPtr) $ \ cursor ->
-        runM (runFail (runReader cursor (runReader bytestring (peekNode >>= unmarshalNode))))
+        (Right <$> runReader (UnmarshalState bytestring cursor) (liftIO (peek rootPtr) >>= unmarshalNode))
+          `catch` (pure . Left . getUnmarshalError)
+
+newtype UnmarshalError = UnmarshalError { getUnmarshalError :: String }
+  deriving (Show)
+
+instance Exception UnmarshalError
+
+data UnmarshalState = UnmarshalState
+  { source :: {-# UNPACK #-} !ByteString
+  , cursor :: {-# UNPACK #-} !(Ptr Cursor)
+  }
+
+type MatchM = ReaderC UnmarshalState IO
+
+newtype Match t = Match
+  { runMatch :: forall a . UnmarshalAnn a => Node -> MatchM (t a)
+  }
```

Comment:
This makes me a bit sad, but it was necessary: quantifying over `sig` and `m` in `runMatch` let us hold the specific stack abstract, but by the same token `ghc` can’t specialize it or inline at all because it has to work _for all_ `Reader` carriers, etc., etc.

Still, we get to use the same API for interacting with it, which is nice.

Meanwhile, `MatchM` has to use the `forall` so that we don’t have to build the `matchers` table anew every time we call into it: constraints are compiled to arguments and so we’d lose the CAFs that the matcher tables otherwise get compiled to.